### DISCUSSION TITLE
Is it possible to use selected venv in python tool, or install some extra libraries?

### DISCUSSION THREAD

**Discussion by johnnieskywalker** [microsoft/promptflow#3565] [Q&A] [answered]
I'm trying to add python tool and import some libraries but I have no possibility to specify virtual environemnt or do pip install, what can I do? Is it possible to install libraries and use them in Python tool?

**Comment by brynn-code** [comment]
Hi, the extra requirements can be specified for flow by `environment: python_requirements_txt`, for example: https://github.com/microsoft/promptflow/blob/main/examples/flows/standard/web-classification/flow.dag.yaml, if you are deploying the flow, this will be installed via dockerfile. If you are testing the flow, or just running the flow locally, you may need to install the requirements manually, as promptflow will not do environment management. 
Document about this key: https://microsoft.github.io/promptflow/reference/flow-yaml-schema-reference.html?highlight=python_requirements_txt

**Reply by johnnieskywalker** [reply]
So it can be done only if developing flow locally in vs code and can't be done in Azure Ml Studio web app in Prompt Flow web editor then?

**Reply by brynn-code** [reply, marked as answer]
If you are developing flow in Azure AI studio, that's the same, just add the extra requirements and specify in the flow.dag.yaml, the requirements.txt file is copied to the flow folder, and use it to start your compute session.  Reach here: https://learn.microsoft.com/en-us/azure/machine-learning/prompt-flow/how-to-manage-compute-session?view=azureml-api-2&tabs=cli#update-a-compute-session-on-the-studio-flow-page
